updated storage snippets

csells opened the pull request.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
     public class StorageClientSnippets
     {
-        private readonly StorageSnippetFixture _fixture;
+        private readonly StorageSnippetFixture fixture;
```
Whilst this is my personal preferred style as well, our code formatter uses `_instanceField` and `s_staticField`.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
 
         [Fact]
-        public void ListObjects()
+        void ListBuckets()
```
I was quite surprised that non-public test methods are still run... but it seems fine. I learn something new every day :)

csells replied:
To which code formatter are you referring? Is it part of the repo or an OSS tool w/ config specified for contributors that I've missed?

jskeet replied:
It's the one linked in

https://github.com/GoogleCloudPlatform/gcloud-dotnet/blob/master/CONVENTIONS.md

But no, I'm afraid we haven't got round to writing up usage explicitly in here yet. Ideally it should be part of the Travis build. More polish to apply...

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+
+            // List all buckets associated with a project
+            var buckets = client.ListBuckets(projectId);
+            // </ListBuckets>
```
Assert that the fixture's bucket is in the list of buckets returned? (Feel free to punt any of these to me...)

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+            var bucketName = Guid.NewGuid().ToString();
+
+            // Bucket defined in Google.Apis.Storage.v1.Data namespace
+            var bucket = client.Service.Buckets.Insert(new Bucket { Name = bucketName }, projectId).Execute();
```
Will look forward to sending you a pull request with this simplified when I've implemented StorageClient.CreateBucket...

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+            }
+            // </DownloadObject>
+
+            // want to show the source path in the snippet to give an idea
```
Ack. Will try to think of a cunning way of avoiding this as a problem if we run into it a lot. For the moment, this seems like a good workaround.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+            // <UploadObject>
+            var client = StorageClient.Create();
+            var source = @".\hello.txt";
+            var destination = "greetings/hello.txt";
```
Possibly write to a different storage object here than in the DownloadFile case?

csells replied:
I thought of that, but I didn't find one I like as well as "hello.txt". That said, picking something different from upload and download removes a bunch of setup and teardown code in the test, so I'll think again.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+
+                // ObjectsResource defined in Google.Apis.Storage.v1 namespace
+                // var acl = ObjectsResource.InsertMediaUpload.PredefinedAclEnum.PublicRead // public
+                var acl = ObjectsResource.InsertMediaUpload.PredefinedAclEnum.AuthenticatedRead; // private
```
I'd expect anyone using C# 6 to have a `using static` statement for this, but obviously we can't easily include that in the snippet. It's a shame to give an "uglier" impression of the code than what a user might actually use, but again I can't think of a better alternative right now. Any thoughts welcome.

csells replied:
I was thinking of "snippet annotations", e.g.

<UploadObject note="using Foo.Bar;" note="...">

which would come out as a snippet like this:

using Foo.Bar;
...
<the code snippet>

jskeet replied:
That would certainly seem like a reasonable output format, but will require more work to implement. (The current extraction is almost entirely from docfx - I just automated attaching the snippets to the prod methods.)

If we decide to divorce ourself from the automated extraction of course, that does mean we could do more cunning processing - like "inlining" constants to avoid the tension between code duplication and being standalone.

csells replied:
Certainly a little duplicate code, with some asserts, is better than sacrificing the readability of the snippet, so my little work-around works for now.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -12,33 +12,212 @@
+            // want to make sure it matches the ones in the fixture
+            Assert.Equal(fixture.HelloWorldName, objectName);
+
+            // need to put the object back as it's used elsewhere
```
I think it would be cleaner for this test to create a different file to start with (before the snippet part) and then delete it.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageSnippetFixture.cs:
```diff
@@ -42,30 +42,30 @@ public StorageSnippetFixture()
 
         /// <summary>
-        /// Creates a new bucket with the ID <see cref="BucketId"/>, and populates it with files needed in
+        /// Creates a new bucket with the ID <see cref="BucketName"/>, and populates it with files needed in
```
Change "with the ID" to "with the name"? (Agreed it's better as `BucketName` though.)

jskeet commented:
LGTM apart from the field name change - other suggestions are optional, although I do prefer the idea of tests not affecting state used by other tests. (So uploading to a separate file, and creating/deleting another file.)

csells commented:
Updated based on feedback and added another Overview test meant to show up as an overview snippet for getting started from scratch scenarios, e.g. on the front page of http://googlecloudplatform.github.io/gcloud-dotnet/.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -27,30 +27,73 @@ namespace Google.Storage.V1.Snippets
+            Assert.Contains(client.ListBuckets(projectId), b => b.Name == bucketName);
+
+            File.Delete(@".\file1.txt");
+            client.DeleteObject(bucketName, "file1.txt");
```
Maybe we should put something in the fixture as a way of registering extra buckets that have been created, so they can be deleted automatically at the end of the test too? (That way stuff will be deleted even if the assertions fail.) Definitely not suggesting for now, but let me know if you like the idea and I'll see what I can do.

csells replied:
Registering buckets, objects and files for clean-up by the fixture would be useful for the case where the assertions fail, as you say. Good idea.

jskeet commented on snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs:
```diff
@@ -95,15 +95,15 @@ public void CreateBucket()
             var projectId = _fixture.ProjectId;
 
-            // <Buckets.Insert>
+            // <Service.Buckets.Insert>
```
Don't worry about this for the moment - basically the snippets are only likely to be meaningful within StorageClient. But might as well leave the commit in place.

csells replied:
I’d like us to find a way to show even the ugly snippets because even they’re WAY better than none at all.


On May 7, 2016, at 9:06 AM, Jon Skeet <notifications@github.com<mailto:notifications@github.com>> wrote:


In snippets/Google.Storage.V1.Snippets/StorageClientSnippets.cs<https://github.com/GoogleCloudPlatform/gcloud-dotnet/pull/141#discussion_r62418274>:

> @@ -95,15 +95,15 @@ public void CreateBucket()
>          {
>              var projectId = _fixture.ProjectId;
>
> -            // <Buckets.Insert>
> +            // <Service.Buckets.Insert>


Don't worry about this for the moment - basically the snippets are only likely to be meaningful within StorageClient. But might as well leave the commit in place.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/GoogleCloudPlatform/gcloud-dotnet/pull/141/files/9b3f60a6a7c089fa672b86b0cc93f81d09178459..6b6522308a2e56861953bea7b5fc4e601db0b969#r62418274>

jskeet replied:
But they'd be snippets on a class which we don't have the documentation for in this repository... there's just nowhere for them to go. If we start including *all* the code in the docs (as per my experiment at http://jskeet.github.io/gcloud-dotnet) then we could potentially do it, but I'll need to think about exactly how.

csells replied:
I’m in favor of including all of the reference docs, but we’ll have to figure out a way to lay them out in layers.

jskeet commented:
LGTM - thanks for the fixups. Will experiment with how to add the overview snippet into the front page. Merge at will.
